vision/diffusion_mnist/diffusion_mnist.jl

```diff
@@ -0,0 +1,330 @@
+
+"""
+Makes the UNet struct callable and shows an example of a "Functional" API for modeling in Flux. \n
+"""
+function (unet::UNet)(x, t)
+    # Embedding
+    embed = unet.layers.gaussfourierproj(t)
+    embed = swish.(unet.layers.linear(embed))
+    # Encoder
+    h1 = unet.layers.conv1(x)
+    h1 = h1 .+ expand_dims(unet.layers.dense1(embed), 2)
+    h1 = unet.layers.gnorm1(h1)
+    h1 = swish.(h1)
+    h2 = unet.layers.conv2(h1)
+    h2 = h2 .+ expand_dims(unet.layers.dense2(embed), 2)
+    h2 = unet.layers.gnorm2(h2)
+    h2 = swish.(h2)
+    h3 = unet.layers.conv3(h2)
+    h3 = h3 .+ expand_dims(unet.layers.dense3(embed), 2)
+    h3 = unet.layers.gnorm3(h3)
+    h3 = swish.(h3)
+    h4 = unet.layers.conv4(h3)
+    h4 = h4 .+ expand_dims(unet.layers.dense4(embed), 2)
+    h4 = unet.layers.gnorm4(h4)
+    h4 = swish.(h4)
+    # Decoder
+    h = unet.layers.tconv4(h4)
+    h = h .+ expand_dims(unet.layers.dense5(embed), 2)
+    h = unet.layers.tgnorm4(h)
+    h = swish.(h)
+    h = unet.layers.tconv3(cat(h, h3; dims=3))
+    h = h .+ expand_dims(unet.layers.dense6(embed), 2)
+    h = unet.layers.tgnorm3(h)
+    h = swish.(h)
+    h = unet.layers.tconv2(cat(h, h2, dims=3))
+    h = h .+ expand_dims(unet.layers.dense7(embed), 2)
+    h = unet.layers.tgnorm2(h)
+    h = swish.(h)
+    h = unet.layers.tconv1(cat(h, h1, dims=3))
+    # Scaling Factor
+    h ./ expand_dims(unet.layers.marginal_prob_std(t), 3)
+end
+
+"""
+Model loss following the denoising score matching objectives:
+
+# Notes
+Denoising score matching objective:
+```julia
+min wrt. θ (
+    𝔼 wrt. 𝘵 ∼ 𝒰(0, 𝘛)[
+        λ(𝘵) * 𝔼 wrt. 𝘹(0) ∼ 𝒫₀(𝘹) [
+            𝔼 wrt. 𝘹(t) ∼ 𝒫₀ₜ(𝘹(𝘵)|𝘹(0)) [
+                (||𝘚₀(𝘹(𝘵), 𝘵) - ∇ log [𝒫₀ₜ(𝘹(𝘵) | 𝘹(0))] ||₂)²
+            ]
+        ]
+    ]
+)
+``` 
+Where 𝒫₀ₜ(𝘹(𝘵) | 𝘹(0)) and λ(𝘵), are available analytically and
+𝘚₀(𝘹(𝘵), 𝘵) is estimated by a U-Net architecture.
+
+# References:
+http://www.iro.umontreal.ca/~vincentp/Publications/smdae_techreport.pdf \n
+https://yang-song.github.io/blog/2021/score/#estimating-the-reverse-sde-with-score-based-models-and-score-matching \n
+https://yang-song.github.io/blog/2019/ssm/
+"""
+function model_loss(model, x, device, ϵ=1.0f-5)
+    batch_size = size(x)[end]
+    # (batch) of random times to approximate 𝔼[⋅] wrt. 𝘪 ∼ 𝒰(0, 𝘛)
+    random_t = rand(Float32, batch_size) .* (1.0f0 - ϵ) .+ ϵ |> device
+    # (batch) of perturbations to approximate 𝔼[⋅] wrt. 𝘹(0) ∼ 𝒫₀(𝘹)
+    z = randn!(similar(x))
+    std = expand_dims(model.layers.marginal_prob_std(random_t), 3)
+    # (batch) of perturbed 𝘹(𝘵)'s to approximate 𝔼 wrt. 𝘹(t) ∼ 𝒫₀ₜ(𝘹(𝘵)|𝘹(0))
+    perturbed_x = x + z .* std
+    # 𝘚₀(𝘹(𝘵), 𝘵)
+    score = model(perturbed_x, random_t)
+    # mean over batches
+    sum(
+        # L₂ norm over WHC dimensions
+        sum(
+            (score .* std + z) .^ 2,
+            dims=(1, 2, 3)
+        )
+    ) / batch_size
```

Comment:
The outer sum can just be a `mean`, no?

Reply:
I couldn't mind a mean without using another import... figured `sum() / length()` was easy enough 🤷‍♂️

Reply:
`Statistics.mean` is part of the stdlib and should be favoured for many reasons: performance, semantic correctness, terseness (we try to avoid breaking nested expressions like this over too many lines), best practices (many new users don't know it exists). I would not think twice about using it :)